File: tree/11-路径总和/solve2.py
# https://leetcode-cn.com/problems/path-sum/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if not root:
            return False
        stack   = [ root, ]
        cur_sum = [ sum, ]
        while len(stack) > 0:
            root = stack.pop()
            sum  = cur_sum.pop()
            if not root.right and not root.left:
                if root.val == sum:
                    return True
            if root.right:
                stack.append(root.right)
                cur_sum.append(sum - root.val)
            if root.left:
                stack.append(root.left)
                cur_sum.append(sum - root.val)
        return False

File: two point/167. Two Sum II - Input array is sorted/solve_2.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int i = 0;
        int j = numbers.size() - 1;
        int sum = 0;
        // 两边向中间遍历
        while(i < j)
        {
            sum = numbers[i] + numbers[j];
            if(target == sum) break;
            else if(target < sum) i++;
            else if(target < sum) j--;
        }
        return vector<int>{i + 1, j + 1};
    }
};